### PR TITLE
Introduce the Cluster class holding the runtime state of the VPA recommender

### PR DESCRIPTION
**Pull request by kgrygiel**
The Cluster object keeps a map with all pods. Each pod keeps a map with its containers.
Additionally each pod keeps information about labels (this will be necessary to handle VPA label selectors).
The Cluster.AddSample method dispatches usage sample to the proper ContainerStats objects.

### CODE REVIEW

File: vertical-pod-autoscaler/recommender/model/cluster.go [hunk 1]
```diff
@@ -0,0 +1,98 @@
...
+	}
+	if _, containerExists := pod.Containers[containerID.ContainerName]; !containerExists {
+		pod.Containers[containerID.ContainerName] = NewContainerStats()
+	}
+	return nil
+}
+
+// AddSample adds a new usage sample to the proper container in the Cluster
+// object. Requires the container as well as the parent pod to be added to the
+// Cluster first. Otherwise an error is returned.
```
Comment (KarolKraskiewicz) [inline vertical-pod-autoscaler/recommender/model/cluster.go]: why not add pods and containers automatically?

File: vertical-pod-autoscaler/recommender/model/cluster.go [hunk 2]
```diff
@@ -0,0 +1,98 @@
...
+
+// Cluster holds all runtime information about resources in the cluster
+// required for VPA operations.
+type Cluster struct {
+	// Pods in the cluster.
+	Pods map[PodID]*Pod
+}
+
+// Pod holds runtime information about a single Pod.
+type Pod struct {
```
Comment (mwielgus) [inline vertical-pod-autoscaler/recommender/model/cluster.go]: There will be a name collision between pod and kubernetes pod. Rename.

File: vertical-pod-autoscaler/recommender/model/cluster.go [hunk 3]
```diff
@@ -0,0 +1,98 @@
...
+func NewCluster() *Cluster {
+	return &Cluster{
+		make(map[PodID]*Pod), // empty pods map.
+	}
+}
+
+// ContainerUsageSampleWithKey holds a ContainerUsageSample together with the
+// ID of the container it belongs to.
+type ContainerUsageSampleWithKey struct {
+	ContainerUsageSample
```
Comment (mwielgus) [inline vertical-pod-autoscaler/recommender/model/cluster.go]: Please resolve the naming with @KarolKraskiewicz. I would like to aviod having multiple structs with names like:
* ContainerUsageSnapshot
* ContainerUtilisationSnapshot
* ContainerUsageSample

File: vertical-pod-autoscaler/recommender/model/cluster.go [hunk 4]
```diff
@@ -0,0 +1,98 @@
...
+
+package model
+
+import (
+	labels "k8s.io/apimachinery/pkg/labels"
+)
+
+// Cluster holds all runtime information about resources in the cluster
+// required for VPA operations.
+type Cluster struct {
```
Comment (mwielgus) [inline vertical-pod-autoscaler/recommender/model/cluster.go]: Can you be more explicit with the name. What info is stored in this struct?

File: vertical-pod-autoscaler/recommender/model/cluster.go [hunk 5]
```diff
@@ -0,0 +1,98 @@
...
+		cluster.Pods[podID] = newPod(podID)
+	}
+	cluster.Pods[podID].Labels = labels
+}
+
+// AddOrUpdateContainer creates a new container with the given ContainerID and
+// adds it to the parent pod in the Cluster object, if not yet present.
+// Requires the pod to be added to the Cluster first. Otherwise an error is
+// returned.
+func (cluster *Cluster) AddOrUpdateContainer(containerID ContainerID) error {
```
Comment (mwielgus) [inline vertical-pod-autoscaler/recommender/model/cluster.go]: Karol is also adding ContainerId.

File: vertical-pod-autoscaler/recommender/model/cluster.go [hunk 6]
```diff
@@ -0,0 +1,98 @@
...
+	}
+	containerStats, containerExists := pod.Containers[sample.Container.ContainerName]
+	if !containerExists {
+		return NewKeyError(sample.Container)
+	}
+	containerStats.AddSample(&sample.ContainerUsageSample)
+	return nil
+}
+
+func newPod(id PodID) *Pod {
```
Comment (mwielgus) [inline vertical-pod-autoscaler/recommender/model/cluster.go]: Why do we need this?

File: vertical-pod-autoscaler/recommender/model/cluster.go (hunk 3 above)
Comment (KarolKraskiewicz) [reply vertical-pod-autoscaler/recommender/model/cluster.go]: I renamed `containerUsageSnapshot` to `containerMetricsSnapshot` but it is just an internal object. 
The only one which is exposed is `ContainerUtilisationSnapshot` which is somehow similar to `ContainerUsageSample` but `ContainerUtilisationSnapshot` contains some more information like:
* duration of the measurement
* docker image name
* pod creation time
* requested resources (in addition to actual usage)

`ContainerUtilisationSnapshot` is also different in a way that it aggregates data from pods and containers into single, flat structure while here we have some hierarchy. I could refactor Metrics client in a way it produces more hierarchical model if I had a better understanding how is it going to be consumed.

File: vertical-pod-autoscaler/recommender/model/cluster.go (hunk 5 above)
Comment (KarolKraskiewicz) [reply vertical-pod-autoscaler/recommender/model/cluster.go]: My containerId It is basically the same information but structured in a different way. I could use either and should be able to refactor my code easily as soon as both changes are merged.

File: vertical-pod-autoscaler/recommender/model/cluster.go (hunk 3 above)
Comment (kgrygiel) [reply vertical-pod-autoscaler/recommender/model/cluster.go]: Done. Our code now shares the same structures, defined in types.go.

**Pull request merged**